src_kt/java/arcs/crdt/entity/FieldValueInterpreter.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.crdt.entity
+
+import arcs.crdt.internal.ReferenceId
+import kotlin.reflect.KClass
+
+/**
+ * Defines an interpreter capable of serializing, deserializing, and identifying the value for an
+ * entity's field.
+ */
+interface FieldValueInterpreter<T> {
+  /** Identifies the [value]. */
+  fun getReferenceId(value: T): ReferenceId
+
+  /** Serializes the [value] to a [String]. */
+  fun serialize(value: T): String
+
+  /** Deserializes the [rawValue] to an instance of [T]. */
+  fun deserialize(rawValue: String): T
+
+  companion object {
+    private val registeredInterpreters = mutableMapOf<KClass<*>, FieldValueInterpreter<*>>()
+
+    /** Registers one or more [FieldValueInterpreter]s for later use. */
+    fun register(vararg interpreters: Pair<KClass<*>, FieldValueInterpreter<*>>) {
+      interpreters.forEach { (kClass, interpreter) ->
+        registeredInterpreters[kClass] = interpreter
+      }
+    }
+
+    /**
+     * Returns whether or not a [FieldValueInterpreter] has been registered for the given [KClass].
+     */
+    fun containsInterpreterFor(kClass: KClass<*>): Boolean = kClass in registeredInterpreters
+
+    /**
+     * Gets the [FieldValueInterpreter] associated with the given [KClass].
+     *
+     * @throws IllegalArgumentException if no [FieldValueInterpreter] has been registered for
+     *   [kClass].
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun <T> requireInterpreter(kClass: KClass<*>): FieldValueInterpreter<T> =
+      requireNotNull(registeredInterpreters[kClass] as? FieldValueInterpreter<T>) {
+        "No FieldValueInterpreter registered for $kClass"
+      }
+
+    /**
+     * Gets a [ReferenceId] for the given value.
+     *
+     * @throws IllegalArgumentException if no [FieldValueInterpreter] has been registered for [T].
+     */
+    inline fun <reified T> getReferenceId(value: T): ReferenceId =
+      requireInterpreter<T>(T::class).getReferenceId(value)
```

Comment:
Totally optional, but I think this would be slightly nicer if callers just got an interpreter instance and called these methods themselves:

```
val interpreter = FieldValueInterpreter.getInterpreter(value)
interpreter.getReferenceId()
interpreter.serialize()
```

Reply:
That's fair, I'll take another pass through all this stuff to tidy up where possible when Crdts are done, and as I work on the actual storage layer.